File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} subRoot
 * @return {boolean}
 */
var isSubtree = function(root, subRoot) {
    
    const isIdentical = (s,t) =>{
        if (!s && !t) return true;
        if (!s || !t) return false;
        return s.val == t.val && isIdentical(s.left, t.left) && isIdentical(s.right, t.right);
    }
    
    const traverse = (node) =>{
        if(!node) return false;
        
        if(isIdentical(node, subRoot)) return true;
        
        return traverse(node.left) || traverse(node.right);
    }
    
    return traverse(root);
};

File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if(root == null || subRoot == null) return false;
        
        if(isSame(root, subRoot)) return true;
        
        return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
    }
    
    public boolean isSame(TreeNode node, TreeNode subRoot){
        if(node == null && subRoot == null) return true;
        if(node == null || subRoot == null) return false;
        
        return node.val == subRoot.val && isSame(node.left, subRoot.left) && isSame(node.right, subRoot.right);
    }
}